### PR TITLE
filesystem: switch file operations to an interface

### PR DESCRIPTION
**Pull request by Velaron**
This should make adding new fs types easier (useful for android), suggestions are welcome

### CODE REVIEW

File: filesystem/dir.c [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	Q_strncpy( dst, search->filename, size );
+}
+
+int FS_FindFile_DIR( searchpath_t *search, const char *path )
+{
+	char netpath[MAX_SYSPATH];
+
+	Q_sprintf( netpath, "%s%s", search->filename, path );
+
+	if( FS_SysFileExists( netpath, !( search->flags & FS_CUSTOM_PATH ) ) )
```
Comment (nekonomicon) [inline filesystem/dir.c:43]: What if use [va](https://github.com/FWGS/xash3d-fwgs/blob/master/public/crtlib.c#L627=https://github.com/FWGS/xash3d-fwgs/blob/master/public/crtlib.c#L627=) here?

File: filesystem/dir.c [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	separator = Q_max( separator, colon );
+	basepathlength = separator ? (separator + 1 - pattern) : 0;
+	basepath = Mem_Calloc( fs_mempool, basepathlength + 1 );
+	if( basepathlength ) memcpy( basepath, pattern, basepathlength );
+	basepath[basepathlength] = 0;
+
+	Q_sprintf( netpath, "%s%s", search->filename, basepath );
+
+	stringlistinit( &dirlist );
+	listdirectory( &dirlist, netpath, caseinsensitive );
```
Comment (nekonomicon) [inline filesystem/dir.c:70]: And here?

File: filesystem/dir.c [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	Mem_Free( basepath );
+}
+
+int FS_FileTime_DIR( struct searchpath_s *search, const char *filename )
+{
+	char path[MAX_SYSPATH];
+
+	// found in the filesystem?
+	Q_sprintf( path, "%s%s", search->filename, filename );
+	return FS_SysFileTime( path );
```
Comment (nekonomicon) [inline filesystem/dir.c:100]: And here?

File: filesystem/dir.c [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+file_t *FS_OpenFile_DIR( struct searchpath_s *search, const char *filename, const char *mode, int pack_ind )
+{
+	char path[MAX_SYSPATH];
+
+	// found in the filesystem?
+	Q_sprintf( path, "%s%s", search->filename, filename );
+	return FS_SysOpen( path, mode );
```
Comment (nekonomicon) [inline filesystem/dir.c:109]: And here?

File: filesystem/dir.c (hunk 1 above)
Comment (Velaron) [reply filesystem/dir.c:43]: Then it stops working, probably 16 string are not enough

File: filesystem/dir.c [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+	basepath = Mem_Calloc( fs_mempool, basepathlength + 1 );
+	if( basepathlength ) memcpy( basepath, pattern, basepathlength );
+	basepath[basepathlength] = '\0';
+
+	Q_strcpy( temp,  basepath );
+
+	stringlistinit( &dirlist );
+	listdirectory( &dirlist, netpath, caseinsensitive );
+
+	Q_strcpy( temp,  basepath );
```
Comment (a1batross) [inline filesystem/dir.c:74]: Why do you `strcpy` to `temp` right after line 69 (nice), though it doesn't seem changed?

File: filesystem/dir.c [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+	for( dirlistindex = 0; dirlistindex < dirlist.numstrings; dirlistindex++ )
+	{
+		Q_strcpy( &temp[basepathlength], dirlist.strings[dirlistindex] );
```
Comment (a1batross) [inline filesystem/dir.c:78]: It's possible to overwrite stack here.

Use Q_strncpy: `Q_strcpy( &temp[i], str, sizeof( temp ) - i )`;

File: filesystem/dir.c [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+	stringlistfreecontents( &dirlist );
+
+	Mem_Free( basepath );
+}
+
+int FS_FileTime_DIR( searchpath_t *search, const char *filename )
+{
+	char path[MAX_SYSPATH];
+
+	Q_sprintf( path, "%s%s", search->filename, filename );
```
Comment (a1batross) [inline filesystem/dir.c:102]: It's possible to overwrite stack here too. Use Q_snprintf:
`Q_snprintf( path, sizeof( path ), ...`

File: filesystem/dir.c [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+	return FS_SysFileTime( path );
+}
+
+file_t *FS_OpenFile_DIR( searchpath_t *search, const char *filename, const char *mode, int pack_ind )
+{
+	char path[MAX_SYSPATH];
+
+	Q_sprintf( path, "%s%s", search->filename, filename );
```
Comment (a1batross) [inline filesystem/dir.c:110]: Ditto

File: filesystem/dir.c [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+void FS_PrintInfo_DIR( searchpath_t *search, char *dst, size_t size )
+{
+	Q_strncpy( dst, search->filename, size );
+}
+
+int FS_FindFile_DIR( searchpath_t *search, const char *path )
+{
+	char netpath[MAX_SYSPATH];
+
+	Q_sprintf( netpath, "%s%s", search->filename, path );
```
Comment (a1batross) [inline filesystem/dir.c:41]: This line too.

File: filesystem/dir.c [hunk 10]
```diff
@@ -57,21 +57,25 @@ void FS_Search_DIR( searchpath_t *search, stringlist_t *list, const char *patter
...
 	separator = Q_max( separator, colon );
+	
 	basepathlength = separator ? (separator + 1 - pattern) : 0;
 	basepath = Mem_Calloc( fs_mempool, basepathlength + 1 );
 	if( basepathlength ) memcpy( basepath, pattern, basepathlength );
-	basepath[basepathlength] = 0;
+	basepath[basepathlength] = '\0';
 
-	Q_sprintf( netpath, "%s%s", search->filename, basepath );
+	Q_strcpy( temp,  basepath );
```
Comment (nekonomicon) [inline filesystem/dir.c]: What are you doing?

Comment (Velaron) [reply filesystem/dir.c]: meth

Comment (Velaron) [reply filesystem/dir.c]: Actually I can't find this code, so not sure where did you get that?

**Review by a1batross — APPROVED** [review]
Looks fine to me. Should test that first.

**Comment by a1batross** [comment]
Was merged already. GitHub didn't detect it lol

**Pull request closed without merging**